astrolabe/spec_runner.py

```diff
@@ -0,0 +1,403 @@
+        assert_subset(
+            cluster_config,
+            self.spec.maintenancePlan[state].clusterConfiguration)
+        process_args = self.cluster_url.processArgs.get().data
+        assert_subset(
+            process_args, self.spec.maintenancePlan[state].processArgs)
+
+    def initialize(self):
+        """
+        Initialize a cluster with the configuration required by the test
+        specification.
+        """
+        __logger__.info("Initializing cluster {!r}".format(self.cluster_name))
+
+        cluster_config = self.spec.maintenancePlan.initial.\
+            clusterConfiguration.copy()
+        cluster_config["name"] = self.cluster_name
+        try:
+            self.client.groups[self.get_group().id].clusters.post(
+                **cluster_config)
+        except AtlasApiError as exc:
+            if exc.error_code == 'DUPLICATE_CLUSTER_NAME':
+                # Cluster already exists. Simply re-configure it.
+                # Cannot send cluster name when updating existing cluster.
+                cluster_config.pop("name")
+                self.client.groups[self.get_group().id].\
+                    clusters[self.cluster_name].patch(**cluster_config)
+
+        # Apply processArgs if provided.
+        process_args = self.spec.maintenancePlan.initial.processArgs
+        if process_args:
+            self.client.groups[self.get_group().id].\
+                clusters[self.cluster_name].processArgs.patch(**process_args)
+
+    def run(self, persist_cluster=False):
+        __logger__.info("Running test {!r} on cluster {!r}".format(
+            self.id, self.cluster_name))
+
+        # Step-0: sanity-check the cluster configuration.
+        self.verify_cluster_configuration_matches("initial")
+
+        # Start the test timer.
+        timer = Timer()
+        timer.start()
+
+        # Step-1: load test data.
+        test_data = self.spec.driverWorkload.get('testData')
+        if test_data:
+            __logger__.info("Loading test data on cluster {!r}".format(
+                self.cluster_name))
+            connection_string = self.get_connection_string()
+            client = MongoClient(connection_string, w="majority")
+            coll = client.get_database(
+                self.spec.driverWorkload.database).get_collection(
+                self.spec.driverWorkload.collection)
+            coll.drop()
+            coll.insert_many(test_data)
+            __logger__.info(
+                "Successfully loaded test data on cluster {!r}".format(
+                    self.cluster_name))
+
+        # Step-2: run driver workload.
+        __logger__.info("Starting workload executor")
+        connection_string = self.get_connection_string()
+        driver_workload = json.dumps(self.spec.driverWorkload)
+        worker_subprocess = subprocess.Popen([
+            sys.executable, self.config.workload_executor, connection_string,
+            driver_workload], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

Comment:
Since we PIPE the output, we need to continuously read the worker_subprocess output (via `worker_subprocess.stdout.read(..)`) or else we risk blocking the child process when the pipe fills up.

See the note in subprocess.wait:
> This will deadlock when using stdout=PIPE or stderr=PIPE and the child process generates enough output to a pipe such that it blocks waiting for the OS pipe buffer to accept more data. Use Popen.communicate() when using pipes to avoid that.

https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait

A simpler option might be to redirect the worker_subprocess output to a file. Then we can open and read the file after the process exits.

Reply:
Just to be clear: there's no risk of deadlock in the current approach. The risk is that the child process can fill up the pipe with output and then it will block until we call worker_subprocess.communicate() down below.

Reply:
Great point. I have opened #9 for this.